### ISSUE TITLE
Custom complex rules?

### ISSUE THREAD

**Issue by JesseLeite** [tekezo/Karabiner-Elements#803]
I absolutely love the new complex rules.  It's so nice to have vi-like `hjkl` arrows again.  Thank you @tekezo 👌 

Anyway, I'm wondering if it's possible to modify this complex rule to post `escape` when `right_command` is pressed alone?

I'm also missing the `left_control+delete` to forward delete on external keyboards, so I was thinking of adding this as a complex rule as well.

Is it possible to import my own complex rule files?  Also if anyone has any advice on how to achieve the above via complex rules?

PS. If there's a better forum for questions like this, I apologize!

**Comment by tolgraven** [comment]
Seems the readme doesn't cover the new syntax fully, but it will point you to the correct file, `~/.config/karabiner/karabiner.json`, and the syntax is easy to figure out just looking at existing/example rules. Then just duplicate segments and change them to do what you want.

**Comment by pkrawczyk89** [comment]
Just wonder if there is any chance to bring a half-qwerty layout to karabiner elements using these custom complex modifications? It's about mirroring the keyboard while keys are pressed together with the spacebar. It requires changing spacebar into a modifier for the alphanumerics.

**Comment by petter-** [comment]
Does these new custom complex rules enable us to create a toggle NumLock rule set to enable and toggle between TK keyboards like CoolerMaster Masterkeys M which mix the nav keys with numpad keys?

**Comment by pablomedok** [comment]
I'm not updating to Sierra (and High Sierra), because of Karabiner. I need to create several complex custom rules that I almost depend on:
[private.xml.zip](https://github.com/tekezo/Karabiner-Elements/files/1127710/private.xml.zip)

I work in MacOS, Windows in RDP (Jump Desktop) and Windows in VM (Parallels) simultaneously. Before Karabiner it was a nightmare, so I unified all modifiers keys to work like Windows. And now I want to convert them to Karabiner Elements.

As far as I understand, Elements now can only import complex rules prepared by you from your site only.

**Comment by tolgraven** [comment]
@pablomedok `karabiner.json` is the new `private.xml`, that's where imported rules end up when enabled in the GUI, and it's where you can compose your own as well.

**Comment by JesseLeite** [comment]
My post escape problem was addressed in #553 [by this custom complex rule](https://github.com/tekezo/Karabiner-Elements/issues/553#issuecomment-313643016) 👌 

Still curious how I can implement `left_control+delete` to forward delete for non macbook keyboards though?

**Comment by allthetime** [comment]
@pablomedok that's not the case at all, you can just edit ./files/complex_modifications_rules_example.json in this project and then build the dmg yourself and they will all show up in the examples and you can just add them yourself. 

On that note, to anyone else who is aware, how hard would it be to make it possible to import local json, I feel like it must be pretty easy because you can already do it with remote json...

**Comment by JesseLeite** [comment]
Pasting the `karabiner://` link into Chrome's address bar tries to google search though.  Whereas if you setup as an HTML link on a page, then clicking it properly triggers karabiner.  Maybe this depends on the browser though?

**Comment by allthetime** [comment]
Are you sure the link is well formed? Using chrome, opera, or safari I am
able to paste links directly and have the intended effect.

**Comment by JesseLeite** [comment]
Hmm, maybe it's a browser setting in Chrome?

```
karabiner://karabiner/assets/complex_modifications/import?url=https://misc.jesseleite.com/karabiner/basic-caps-vi-mode.json
```

When I paste that in Chrome it tries to perform a google search, but when I paste it in Safari it asks to open Karabiner.

**Comment by allthetime** [comment]
You need to URL encode the url, look at my example again.

use javascript\node console:

encodeURIComponent('https://misc.jesseleite.com/karabiner/basic-caps-vi-
mode.json')

so

https%3A%2F%2Fmisc.jesseleite.com%2Fkarabiner%2Fbasic-caps-vi-mode.json

and

karabiner://karabiner/assets/complex_modifications/import?url=https%3A%2F%2Fmisc.jesseleite.com%2Fkarabiner%2Fbasic-caps-vi-mode.json

**Comment by pablomedok** [comment]
thanks, i'll try

**Comment by GuybrushZ** [comment]
Sorry for the very Newbie question, but I am finding it hard to follow the discussion. I want to create a complex rule to map Caps Lock to the following Shift Command M, and have that setup in Outlook only. How do I do that? I am not able to find a way to create a custom complex rule or way to edit an existing one.

**Comment by pimlottc** [comment]
@pkrawczyk89 I had the same thought about implementing a half-keyboard with Karabiner Elements.  Check out what [Crackle](https://medium.com/@jack_21924/crackle-keyboard-layout-for-special-characters-e4dd04838231) did; it's a layout that adds two new meta keys via complex rules.

You can see the config file for here: [karabiner.json](https://drive.google.com/file/d/0B-vJmiiF8fCkTVBhbzl5MkFhSEE/view).  Basically, there's a variable used as a flag for each modifier key that's checked in key binding.  You could use the same technique for a half-keyboard, it's just a matter of changing each of the individual key bindings for every key and using a different modifier.

**Comment by macitpros** [comment]
I'm just getting started with Karabiner-Elements. Coming from Karabiner I had a mapping that worked great to change the keys used with a Logitech keyboard to switch from mirroring to extended display. Same result you would get with control+f1 on a standard Mac keyboard which I remapped it to f7.

**Comment by JesseLeite** [comment]
I just [wrote up a post about my finished complex modification](https://jesseleite.com/posts/1/vim-all-the-things) here.  Keep the great discussion going, but I'm going to close my issue because my original issue is solved 👍

**Comment by rhuitl** [comment]
@pkrawczyk89 @pimlottc I implemented a half-QWERTZ layout for KE and opened a PR: https://github.com/pqrs-org/KE-complex_modifications/pull/251. Once accepted and published, it can be installed from the UI.

To test custom complex rules, take a look at https://github.com/pqrs-org/KE-complex_modifications#local-testing

**Comment by ihtoyhero** [comment]
I'm trying to map left-control+left-click to just left click, how do I map that? or can someone map one for me so I can import it? thanks! I have a Mac Sierra 10.

**Comment by esseti** [comment]
Can someone explain how I can create a complex rule?  (i need to map FN to command+option)

**Comment by limeb** [comment]
@allthetime is there a way to create a link in HTML file with out creating a separate file in addition to the text in the HTML file?

**Comment by brandondrew** [comment]
@esseti  you want to add your own JSON files to `~/.config/karabiner/assets/complex_modifications/`.  If you look at the examples already in that directory (if you have installed any through the GUI) you'll see how it's done.  After adding your JSON file there, you'll need to enabled it in the GUI.
